### ISSUE TITLE
tf GaussianMLP* primitives seem to be unstable for large hidden sizes in PPO

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/garage#255]
std_share_network=True,
hidden_sizes=(200,200)

observed behavior: std explodes

**Comment by ryanjulian** [comment]
This is almost certainly due to the way we initialize the std_dev of the std_network

**Comment by ryanjulian** [comment]
Please let me know if you are able to reproduce.

**Comment by ryanjulian** [comment]
@CatherineSue did you do any research on this? were you able to reproduce?

Perhaps @ahtsan can try adding this to his unit test?

**Comment by CatherineSue** [comment]
I haven't reproduced this yet. It would be great if @ahtsan can add this to his unit test. I'll be focusing on the policy_entropy bug.

**Comment by ahtsan** [comment]
I run ppo on Pendulum-v0 with GaussianMLP hidden sizes (256, 128, 64) and plot the average std
![screenshot from 2018-10-25 15-15-20](https://user-images.githubusercontent.com/10270515/47533507-d4038900-d868-11e8-9a5d-43e0f01ef319.png)
I think it is exploding

**Comment by ryanjulian** [comment]
Is policy_entropy_coeff > 0

**Comment by ahtsan** [comment]
yes

**Comment by ahtsan** [comment]
turns out `policy_ent_coeff` is already 0 and environment is already normalized
```python
        env = TfEnv(normalize(gym.make("Pendulum-v0")))
        policy = GaussianMLPPolicy(
            env_spec=env.spec,
            hidden_sizes=(256, 128, 64),
            hidden_nonlinearity=tf.nn.tanh,
            output_nonlinearity=None,
            std_share_network=True
        )
        baseline = GaussianMLPBaseline(
            env_spec=env.spec,
            regressor_args=dict(hidden_sizes=(32, 32)),
        )
        algo = PPO(
            env=env,
            policy=policy,
            baseline=baseline,
            batch_size=1024,
            max_path_length=100,
            n_itr=1000,
            discount=0.99,
            gae_lambda=0.98,
            policy_ent_coeff=0.0,
            plot=False,
        )
```
* `policy_entropy_coeff=0.0`
* `adaptive_std=False` (default)
* `std_share_network=True`
* `learn_std=True` (default)
* `use_softplus_entropy=True` (default)

I am trying on TRPO now

**Comment by ahtsan** [comment]
The issue seems not happening on TRPO
```python
        env = TfEnv(normalize(gym.make("Pendulum-v0")))
        policy = GaussianMLPPolicy(
            env_spec=env.spec,
            hidden_sizes=(256, 128,64),
            hidden_nonlinearity=tf.nn.tanh,
            output_nonlinearity=None,
            std_share_network=True
        )
        baseline = GaussianMLPBaseline(
            env_spec=env.spec,
            regressor_args=dict(hidden_sizes=(32, 32)),
        )
        algo = TRPO(
            env=env,
            policy=policy,
            baseline=baseline,
            batch_size=1024,
            max_path_length=100,
            n_itr=1000,
            discount=0.99,
            gae_lambda=0.98,
            policy_ent_coeff=0.0,
            plot=False,
        )
```
* `policy_entropy_coeff=0.0`
* `adaptive_std=False` (default)
* `std_share_network=True`
* `learn_std=True` (default)
* `use_softplus_entropy=True` (default)

<img width="1046" alt="screen shot 2018-10-27 at 9 18 40 am" src="https://user-images.githubusercontent.com/10270515/47608076-8a887a80-d9dd-11e8-8474-90a59c7c2da2.png">

difference is trpo uses vanilla pg loss while ppo uses clip

**Comment by ryanjulian** [comment]
Interesting! @CatherineSue I think we might want to set a max_grad_norm in PPO. I think the openai baselines implementation has something like that.

**Comment by zhanpenghe** [comment]
@ahtsan I ran your experiment and I am pretty sure that the behavior of your experiment is caused by a bad baseline instead of some gradient issues. I have not located the problem yet but I think your parameters are already bad for pendulum environment so it is not actually reproducing the problem. @ryanjulian I am not very sure that if exploding std is an issue of our implementation. It appears for some environments instead of all of them.

**Comment by ryanjulian** [comment]
@zhanpenghe 
even if only happens under some environments, you can't blame the environment. If the algorithm fails under some environments, it's the algorithm's failure. I'm speculating that the gradient clipping is there to avoid this problem.

are you saying you can't reproduce @ahtsan 's failure?

**Comment by zhanpenghe** [comment]
I reproduced it but it seems that it is just a bad set of parameters for pendulum and the policy is not learning at all. I think a good way to actually locate the issue is deriving a good set of parameter of an environment first and then scaling up the policy/baseline size. 

I have not actually reproduced the large hidden size issue.

**Comment by zhanpenghe** [comment]
I tried to used a very large hidden size (512, 256, 64) of inverted-pendulum and it is not exploding.

**Comment by CatherineSue** [comment]
I clipped gradient in the optimizer. But it doesn't work for the pendulum env. I agree with Zhanpeng, looks like using the parameters above for pendulum is not working. The policy is not learning.

**Comment by ryanjulian** [comment]
What do you all suggest we do with this for the release? Just leave it as-is?

**Comment by CatherineSue** [comment]
We can wait for the result of using max_grad_norm. I'm still running the test.

**Comment by CatherineSue** [comment]
![screenshot from 2018-10-30 16-55-17](https://user-images.githubusercontent.com/8605658/47757672-ab163600-dc64-11e8-81ed-513a2dd5c7f4.png)
This is not exploding as much as not using max_grad_norm, but the variance seems high

**Comment by CatherineSue** [comment]
Let's leave this issue for another day. 
Based on my research, clip gradient doesn't seem to solve the problem. And I also have some doubt regarding the parameter setting for now. For example, the baseline hidden size is (64, 64). 
I haven't tried to reproduce this in Inverted-Pendulum because it takes a long time to run a test with such a big network.

**Comment by ryanjulian** [comment]
This appears to be a consequence of PPO rather than a software flaw, so I'm removing the bug label.
